libs/stats/odc/stats/io.py

```diff
@@ -242,5 +284,156 @@ def dump(self, task: Task, ds: Dataset, aux: Optional[Dataset] = None) -> Delaye
+
+        meta = dataset_assembler.to_dataset_doc()
+        # already add all information to dataset_assembler, now convert to odc and stac metadata format
+
+        stac_meta = self.get_eo3_stac_meta(task, meta, stac_file_path, odc_file_path)
+
+        odc_meta_stream = io.StringIO("") # too short, not worth to move to another method.
+        serialise.to_stream(odc_meta_stream, meta)
+        odc_meta = odc_meta_stream.getvalue() # odc_meta is Python str
+
+        proc_info_meta_stream = io.StringIO("")
+        serialise._init_yaml().dump({**dataset_assembler._user_metadata, "software_versions": dataset_assembler._software_versions}, proc_info_meta_stream)
+        proc_info_meta = proc_info_meta_stream.getvalue()
+
+        # fake write result for metadata output, we want metadata file to be
+        # the last file written, so need to delay it until after sha1 files is
+        # written.
+        stac_meta_sha1 = dask.delayed(WriteResult(stac_file_path, mk_sha1(stac_meta), None))
+        odc_meta_sha1 = dask.delayed(WriteResult(odc_file_path, mk_sha1(odc_meta), None))
+        proc_info_sha1 = dask.delayed(WriteResult(proc_info_url, mk_sha1(proc_info_meta), None))
+
+        paths = task.paths("absolute", ext=self._band_ext)
+        cogs = self._ds_to_cog(ds, paths)
+
+        if aux is not None:
+            aux_paths = {
+                k: task.aux_path(k, relative_to="absolute", ext=self._band_ext)
+                for k in aux.data_vars
+            }
+            cogs.extend(self._ds_to_cog(aux, aux_paths))
+
+        thumbnail_cogs = []
+
+        input_geobox = GridSpec(shape=task.geobox.shape, 
+                                transform=task.geobox.transform, 
+                                crs=CRS.from_epsg(task.geobox.crs.to_epsg()))
+
+        # if we do not define the r/g/b values, we will grab the first avaialble variable shape to 
+        # generate full zero numpy
+        zero_band = numpy.zeros_like(ds[list(ds.keys())[0]].values.reshape([task.geobox.shape[0], 
+                                                                            task.geobox.shape[1]]))
+
+        if task.product.preview_image_singleband:
+            # single_band: {"measurement": "count_clear", "lookup_table": {"0":[150,150,110]}}, or
+            # single_band: {"measurement":'frequency', 'bit':1, 'display_bands': ['blue']} or 
+            # single_band: {"measurement":'frequency', 'bit':1, 'display_bands': ['red', 'green', 'blue']}
+            for single_band in task.product.preview_image_singleband:
+                band = single_band['measurement']
+                thumbnail_path = odc_file_path.split('.')[0] + f"_{band}_thumbnail.jpg"
+
+                tuning_pixels, stretch = FileWrite()._filter_singleband_data(data=ds[band].values.reshape([task.geobox.shape[0], 
+                                                                                                           task.geobox.shape[1]]),
```

Comment:
how is this moving "thumbnail-sized only data", this is native resolution band data, is this done for every output band?

Reply:
You suggested we change the resolution (then reduce the data transmit cost) before pass to next methods? 

In my design, this 'tunning' should only happen to the 'thumbnail_display' bands. E.g. I plan to use WOfS summary frequency as the thumbnail_display band, so I just tuning it and display it it the JEPG 'red, green, blue' layer.